include/ts/apidefs.h.in

```diff
@@ -261,25 +377,48 @@ typedef enum {
     TS_HTTP_LAST_HOOK _must_ be the last element. Only right place
     to insert a new element is just before TS_HTTP_LAST_HOOK.
 
+    @note The TS_HTTP hooks below have to be in the same order as their
+    corresponding TS_EVENT counterparts.  We use this in calculating the
+    corresponding event from a hook ID by summing
+    TS_EVENT_HTTP_READ_REQUEST_HDR with the hook ID (see the invoke call in
+    HttpSM::state_api_callout). For example, the following expression must
+    be true:
+
+    TS_EVENT_HTTP_TXN_CLOSE == TS_EVENT_HTTP_READ_REQUEST_HDR + TS_HTTP_TXN_CLOSE_HOOK
+
  */
 typedef enum {
-  TS_HTTP_READ_REQUEST_HDR_HOOK,
-  TS_HTTP_OS_DNS_HOOK,
-  TS_HTTP_SEND_REQUEST_HDR_HOOK,
-  TS_HTTP_READ_CACHE_HDR_HOOK,
-  TS_HTTP_READ_RESPONSE_HDR_HOOK,
-  TS_HTTP_SEND_RESPONSE_HDR_HOOK,
-  TS_HTTP_REQUEST_TRANSFORM_HOOK,
-  TS_HTTP_RESPONSE_TRANSFORM_HOOK,
-  TS_HTTP_SELECT_ALT_HOOK,
-  TS_HTTP_TXN_START_HOOK,
-  TS_HTTP_TXN_CLOSE_HOOK,
-  TS_HTTP_SSN_START_HOOK,
-  TS_HTTP_SSN_CLOSE_HOOK,
-  TS_HTTP_CACHE_LOOKUP_COMPLETE_HOOK,
-  TS_HTTP_PRE_REMAP_HOOK,
-  TS_HTTP_POST_REMAP_HOOK,
-  TS_HTTP_RESPONSE_CLIENT_HOOK,
+/*
+   The following macro helps maintain the relationship between the TS_HTTP
+   hooks and their TS_EVENT_HTTP counterparts as described in the above
+   doxygen comment.
+*/
+#define REBASE(COMMON) TS_HTTP_##COMMON##_HOOK = TS_EVENT_HTTP_##COMMON - TS_EVENT_HTTP_READ_REQUEST_HDR
+  REBASE(READ_REQUEST_HDR),             // TS_HTTP_READ_REQUEST_HDR_HOOK
+  REBASE(OS_DNS),                       // TS_HTTP_OS_DNS_HOOK
+  REBASE(SEND_REQUEST_HDR),             // TS_HTTP_SEND_REQUEST_HDR_HOOK
+  REBASE(READ_CACHE_HDR),               // TS_HTTP_READ_CACHE_HDR_HOOK
+  REBASE(READ_RESPONSE_HDR),            // TS_HTTP_READ_RESPONSE_HDR_HOOK
+  REBASE(SEND_RESPONSE_HDR),            // TS_HTTP_SEND_RESPONSE_HDR_HOOK
+  REBASE(REQUEST_TRANSFORM),            // TS_HTTP_REQUEST_TRANSFORM_HOOK
+  REBASE(RESPONSE_TRANSFORM),           // TS_HTTP_RESPONSE_TRANSFORM_HOOK
+  REBASE(SELECT_ALT),                   // TS_HTTP_SELECT_ALT_HOOK
+  REBASE(TXN_START),                    // TS_HTTP_TXN_START_HOOK
+  REBASE(TXN_CLOSE),                    // TS_HTTP_TXN_CLOSE_HOOK
+  REBASE(SSN_START),                    // TS_HTTP_SSN_START_HOOK
+  REBASE(SSN_CLOSE),                    // TS_HTTP_SSN_CLOSE_HOOK
+  REBASE(CACHE_LOOKUP_COMPLETE),        // TS_HTTP_CACHE_LOOKUP_COMPLETE_HOOK
+  REBASE(PRE_REMAP),                    // TS_HTTP_PRE_REMAP_HOOK
+  REBASE(POST_REMAP),                   // TS_HTTP_POST_REMAP_HOOK
+  REBASE(REQUEST_BUFFER_READ_COMPLETE), // TS_HTTP_REQUEST_BUFFER_READ_COMPLETE_HOOK
```

Comment:
I wasn't familiar with this hook, so I looked it up in the docs, and it isn't there.  Would be a great thing to add to the docs.  Perhaps in a separate PR.